feat: implement IPLD reachability analysis

Stebalien opened the pull request:
Implements FIP-0071

Fixes #723
Yeets #503 into the sun.

Stebalien commented:
NOTE: We may need to make this PR against the `next` branch... although it shouldn't change behavior or significantly change runtimes so... we may not.

Stebalien commented on fvm/src/kernel/cbor.rs:
```diff
@@ -0,0 +1,198 @@
+                                    .into(),
+                            );
+                    } else if cid.hash().code() == fvm_shared::IDENTITY_HASH {
+                        if cid.codec() == DAG_CBOR {
```
It's enforced by the `Cid` type itself. I'll document that.

Stebalien commented on sdk/src/ipld.rs:
```diff
@@ -11,6 +13,20 @@ pub const UNIT: u32 = sys::ipld::UNIT;
 /// Store a block. The block will only be persisted in the state-tree if the CID is "linked in" to
 /// the actor's state-tree before the end of the current invocation.
 pub fn put(mh_code: u64, mh_size: u32, codec: u64, data: &[u8]) -> SyscallResult<Cid> {
+    // Short-circuit for identity hashes.
```
With the "identity" hash, the data is _inlined_ into the CID itself. 

The inverse is:

https://github.com/filecoin-project/ref-fvm/blob/358772903261ac2ff9c222a5cc9a8f247a9b047e/sdk/src/ipld.rs#L47-L51

Stebalien replied:
Documented.

fridrik01 commented on fvm/src/kernel/default.rs:
```diff
@@ -265,36 +282,51 @@ where
             // TODO: This is really "super fatal". It means we failed to store state, and should
             // probably abort the entire block.
             .or_fatal()?
-            .ok_or_else(|| anyhow!("missing state: {}", cid))
+            .ok_or_else(|| anyhow!("missing reachable state: {}", cid))
             // Missing state is a fatal error because it means we have a bug. Once we do
             // reachability checking (for user actors) we won't get here unless the block is known
             // to be in the state-tree.
             .or_fatal()?;
```
This dance of checking `or_fatal()`, `ok_or_else()` and then `or_fatal()` again seems a bit redundant to me

Stebalien replied:
Yeah, fair enough. I've collapsed it.

fridrik01 reviewed (approved):
I added some minor nits/questions but otherwise this LGTM

Stebalien commented on fvm/src/kernel/blocks.rs:
```diff
@@ -24,47 +25,50 @@ pub type BlockId = u32;
 
 #[derive(Debug, Clone)]
-pub struct Block {
+pub struct Block(Rc<BlockInner>);
```
So we can put it inside the `Rc` (reference counted smart pointer). Othwerwise, the user has to care about the Rc, which would be kind of annoying.

Stebalien commented:
Note: still holding off on merging till I have some actual tests.

Stebalien commented:
To check the benchmarks:

1. Install graphviz.
2. Run `make` in `testing/calibration`.
3. Check the `OnScanIpldLinks`, `OnScanCborFields`, `OnCheckLinks`, and `OnTrackLinks` benchmarks in `testing/calibration/measurements/out/charts/charges/`.

Note: these benchmarks are a little noisy.

Stebalien commented:
@fridrik01 I've added gas costs, tests, and benchmarks. Since you last reviewed, I also moved some code around to match the FIP. Specifically, I refactored the link scanning algorithm to make it easier to support additional codecs in the future.

fridrik01 commented on fvm/src/kernel/blocks.rs:
```diff
@@ -24,47 +25,50 @@ pub type BlockId = u32;
-    // copy the bytes. So we accept some indirection for reliable performance.
-    #[allow(clippy::redundant_allocation)]
-    data: Rc<Box<[u8]>>,
+    data: Box<[u8]>,
+    links: Box<[Cid]>,
```
Ok, you moved the RC pointer into the whole inner struct, I guess the alternative would be to have two RC pointers in this inner block but I guess this is _better_ and would be the same performance

Stebalien replied:
This should have better performance as it leads to a single allocation instead of 2. But yeah, that's why I did this.

Well, 3 allocations instead of 4.

Stebalien commented on fvm/src/ipld/cbor.rs:
```diff
@@ -0,0 +1,137 @@
+    Ok((maj, val))
+}
+
+/// Given a CBOR serialized IPLD buffer, read through all of it and return all "reachable" links.
```
Fixed.

fridrik01 reviewed (approved):
LGTM
